src/box/tuple_compare.cc

```diff
@@ -1518,6 +1540,21 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
 #define HINT_VALUE_DOUBLE_MAX	(exp2(HINT_VALUE_BITS - 1) - 1)
 #define HINT_VALUE_DOUBLE_MIN	(-exp2(HINT_VALUE_BITS - 1))
 
+/**
+ * We need to squeeze 64 bits of seconds and 32 bits of nanoseconds
+ * into 60 bits of hint value. The idea is to represent wide enough
+ * years range, and leave the rest of bits occupied from nanoseconds part:
+ * - 36 bits is enough for time range of [208BC..4147]
+ * - for nanoseconds there is left 24 bits, which are MSB part of
+ *   32-bit value
+ */
+#define HINT_VALUE_SECS_BITS	36
+#define HINT_VALUE_NSEC_BITS	(HINT_VALUE_BITS - HINT_VALUE_SECS_BITS)
+#define HINT_VALUE_SECS_MAX	((1LL << (HINT_VALUE_SECS_BITS - 1)) - 1)
+#define HINT_VALUE_SECS_MIN	(-(1LL << (HINT_VALUE_SECS_BITS - 1)))
+#define HINT_VALUE_NSEC_SHIFT	(sizeof(int) * CHAR_BIT - HINT_VALUE_NSEC_BITS)
```

Comment:
I used to be `int` in the past :) 

[It's not a problem today though, because `int` is 32-bit on amd64 and arm64, thus alias to `int32_t`. But, yes, agreed we should use here whichever type is of particular structure field.]

Reply:
Corrected mentioned type in `sizeof` expression. Also has corrected commentary about supported years range in hints - it was slightly incorrect:
```diff
--- a/src/box/tuple_compare.cc
+++ b/src/box/tuple_compare.cc
@@ -1544,7 +1544,7 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
  * We need to squeeze 64 bits of seconds and 32 bits of nanoseconds
  * into 60 bits of hint value. The idea is to represent wide enough
  * years range, and leave the rest of bits occupied from nanoseconds part:
- * - 36 bits is enough for time range of [208BC..4147]
+ * - 36 bits is enough for time range of [-208-05-13..6325-04-08]
  * - for nanoseconds there is left 24 bits, which are MSB part of
  *   32-bit value
  */
@@ -1552,7 +1552,7 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
 #define HINT_VALUE_NSEC_BITS   (HINT_VALUE_BITS - HINT_VALUE_SECS_BITS)
 #define HINT_VALUE_SECS_MAX    ((1LL << (HINT_VALUE_SECS_BITS - 1)) - 1)
 #define HINT_VALUE_SECS_MIN    (-(1LL << (HINT_VALUE_SECS_BITS - 1)))
-#define HINT_VALUE_NSEC_SHIFT  (sizeof(int) * CHAR_BIT - HINT_VALUE_NSEC_BITS)
+#define HINT_VALUE_NSEC_SHIFT  (sizeof(int32_t) * CHAR_BIT - HINT_VALUE_NSEC_BITS)
 #define HINT_VALUE_NSEC_MAX    ((1ULL << HINT_VALUE_NSEC_BITS) - 1)

 /*
 ```
 P.S.

<details>
  <summary> Here is how to check exact range for 36-bit dates:</summary>

```
tarantool> bit.lshift(1LL, 36) - 1
---
- 68719476735
...

tarantool> -bit.lshift(1LL, 36)
---
- -68719476736
...

tarantool> date = require 'datetime'
---
...

tarantool> T = date.new()
---
...

tarantool> T.epoch = -68719476736
---
...

tarantool> T
---
- -208-05-13T16:27:44Z
...

tarantool> T.epoch = 68719476735 - -68719476736
---
...

tarantool> T
---
- 6325-04-08T15:04:31Z
...
```  
</details>